File: checkIns/src/main/java/com/popcorn/checkIns/outbox/OutboxEvent.java
package com.popcorn.checkIns.outbox;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(
    name = "outbox_events",
    schema = "\"checkIns\"",
    indexes = {
        @Index(name = "idx_outbox_created_at", columnList = "created_at"),
        @Index(name = "idx_outbox_aggregate", columnList = "aggregate_type, aggregate_id"),
        @Index(name = "idx_outbox_event_type", columnList = "event_type"),
        @Index(name = "idx_outbox_partition_key", columnList = "partition_key")
    },
    uniqueConstraints = {
        @UniqueConstraint(name = "uq_outbox_event_id", columnNames = "event_id")
    }
)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class OutboxEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "event_id", nullable = false, updatable = false)
    private UUID eventId;

    @Column(name = "aggregate_type", nullable = false, updatable = false, length = 100)
    private String aggregateType;

    @Column(name = "aggregate_id", nullable = false, updatable = false, length = 255)
    private String aggregateId;

    @Column(name = "event_type", nullable = false, updatable = false, length = 100)
    private String eventType;

    @Column(name = "partition_key", nullable = false, updatable = false, length = 255)
    private String partitionKey;

    @Column(name = "schema_version", nullable = false, updatable = false)
    private Integer schemaVersion;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "event_data", nullable = false, updatable = false, columnDefinition = "jsonb")
    private Map<String, Object> eventData;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "headers", updatable = false, columnDefinition = "jsonb")
    private Map<String, Object> headers;

    @Column(name = "occurred_at", nullable = false, updatable = false)
    private Instant occurredAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "popup_id")
    private UUID popupId;

    @Column(name = "order_goods_id")
    private UUID orderGoodsId;

    @Builder
    private OutboxEvent(
        UUID eventId,
        String aggregateType,
        String aggregateId,
        String eventType,
        String partitionKey,
        Integer schemaVersion,
        Map<String, Object> eventData,
        Map<String, Object> headers,
        Instant occurredAt,
        UUID popupId,
        UUID orderGoodsId
    ) {
        this.eventId = (eventId != null) ? eventId : UUID.randomUUID();
        this.aggregateType = aggregateType;
        this.aggregateId = aggregateId;
        this.eventType = eventType;
        this.partitionKey = partitionKey;
        this.schemaVersion = (schemaVersion != null) ? schemaVersion : 1;
        this.eventData = eventData;
        this.headers = headers;
        this.occurredAt = (occurredAt != null) ? occurredAt : Instant.now();
        this.createdAt = Instant.now();
        this.popupId = popupId;
        this.orderGoodsId = orderGoodsId;
    }

    public static OutboxEvent of(
        String aggregateType,
        String aggregateId,
        String eventType,
        Map<String, Object> payload,
        Map<String, Object> headers
    ) {
        return OutboxEvent.builder()
            .aggregateType(aggregateType)
            .aggregateId(aggregateId)
            .eventType(eventType)
            .partitionKey(aggregateId)
            .eventData(payload)
            .headers(headers)
            .build();
    }

    public static OutboxEvent of(
        String aggregateType,
        String aggregateId,
        String eventType,
        Map<String, Object> payload,
        Map<String, Object> headers,
        UUID popupId,
        UUID orderGoodsId
    ) {
        return OutboxEvent.builder()
            .aggregateType(aggregateType)
            .aggregateId(aggregateId)
            .eventType(eventType)
            .partitionKey(aggregateId)
            .eventData(payload)
            .headers(headers)
            .popupId(popupId)
            .orderGoodsId(orderGoodsId)
            .build();
    }
}

File: payment/src/main/kotlin/com/popcorn/payment/outbox/OutboxEvent.kt
package com.popcorn.payment.outbox

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import java.time.Instant
import java.util.UUID
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes

@Entity
@Table(
    name = "outbox_events",
    indexes = [
        Index(name = "idx_outbox_created_at", columnList = "created_at"),
        Index(name = "idx_outbox_aggregate", columnList = "aggregate_type, aggregate_id"),
        Index(name = "idx_outbox_event_type", columnList = "event_type"),
        Index(name = "idx_outbox_partition_key", columnList = "partition_key")
    ],
    uniqueConstraints = [
        UniqueConstraint(name = "uq_outbox_event_id", columnNames = ["event_id"])
    ]
)
class OutboxEvent(
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @field:Column(name = "event_id", nullable = false, updatable = false)
    val eventId: UUID,

    @field:Column(name = "aggregate_type", nullable = false, updatable = false, length = 100)
    val aggregateType: String,

    @field:Column(name = "aggregate_id", nullable = false, updatable = false, length = 255)
    val aggregateId: String,

    @field:Column(name = "event_type", nullable = false, updatable = false, length = 100)
    val eventType: String,

    @field:Column(name = "partition_key", nullable = false, updatable = false, length = 255)
    val partitionKey: String,

    @field:Column(name = "schema_version", nullable = false, updatable = false)
    val schemaVersion: Int = 1,

    @field:JdbcTypeCode(SqlTypes.JSON)
    @field:Column(name = "event_data", nullable = false, updatable = false, columnDefinition = "jsonb")
    val eventData: Map<String, Any?>,

    @field:JdbcTypeCode(SqlTypes.JSON)
    @field:Column(name = "headers", updatable = false, columnDefinition = "jsonb")
    val headers: Map<String, Any?>? = null,

    @field:Column(name = "occurred_at", nullable = false, updatable = false)
    val occurredAt: Instant = Instant.now(),

    @field:Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: Instant = Instant.now()
)
